src/hotspot/share/runtime/continuationFreezeThaw.cpp

```diff
@@ -1678,6 +1722,108 @@ bool FreezeBase::check_valid_fast_path() {
   }
   return true;
 }
+
+static void verify_frame_kind(frame& top, Continuation::preempt_kind preempt_kind, Method** m_ptr, const char** code_name_ptr, int* bci_ptr, stackChunkOop chunk) {
+  Method* m;
+  const char* code_name;
+  int bci;
+  if (preempt_kind == Continuation::monitorenter) {
+    assert(top.is_interpreted_frame() || top.is_runtime_frame(), "");
```

Comment:
So could you use precond() here since it's a precondition and you have no assert message?

Reply:
Yes, but don’t really see the benefit. It’s replacing a null string for `precond` in a crash.

Reply:
So is it a compiled frame otherwise? Reporting the unexpected frame type might be useful.

Reply:
I added a check for compiled or native. If we want to check all possibilities we could add a method in frame class to return the name of the frame type.